### PR TITLE
rpc: Document default values for optional arguments

### PR DESCRIPTION
**Pull request by maflcko**

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -457,11 +457,11 @@ static UniValue createrawtransaction(const JSONRPCRequest& request)
                 {
                     {"inputs", RPCArg::Type::ARR, /* opt */ false, /* default_val */ "", "A json array of json objects",
                         {
-                            {"", RPCArg::Type::OBJ, /* opt */ false, /* default_val */ "", "",
+                            {"", RPCArg::Type::OBJ, /* opt */ true, /* default_val */ "", "",
                                 {
                                     {"txid", RPCArg::Type::STR_HEX, /* opt */ false, /* default_val */ "", "The transaction id"},
                                     {"vout", RPCArg::Type::NUM, /* opt */ false, /* default_val */ "", "The output number"},
-                                    {"sequence", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "", "The sequence number"},
+                                    {"sequence", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "depends on the value of the 'replacable' and 'locktime' arguments", "The sequence number"},
```
Comment (practicalswift) [inline src/rpc/rawtransaction.cpp]: Should be "replaceable" :-)

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -1603,7 +1603,7 @@ UniValue createpsbt(const JSONRPCRequest& request)
...
```
Comment (practicalswift) [inline src/rpc/rawtransaction.cpp]: Same here – should be "replaceable" :-)

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -519,7 +519,7 @@ static UniValue decoderawtransaction(const JSONRPCRequest& request)
                 "\nReturn a JSON object representing the serialized, hex-encoded transaction.\n",
                 {
                     {"hexstring", RPCArg::Type::STR_HEX, /* opt */ false, /* default_val */ "", "The transaction hex string"},
-                    {"iswitness", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "", "Whether the transaction hex is a serialized witness transaction\n"
+                    {"iswitness", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "depends on heuritic tests", "Whether the transaction hex is a serialized witness transaction\n"
```
Comment (practicalswift) [inline src/rpc/rawtransaction.cpp]: Should be "heuristic" :-)

**Review by practicalswift — CHANGES_REQUESTED** [review]

**Comment by MarcoFalke** [comment]
Thanks @practicalswift. Fixed my language

File: src/rpc/mining.cpp [hunk 4]
```diff
@@ -888,9 +888,9 @@ static UniValue estimaterawfee(const JSONRPCRequest& request)
                 "defined in BIP 141 (witness data is discounted).\n",
                 {
                     {"conf_target", RPCArg::Type::NUM, /* opt */ false, /* default_val */ "", "Confirmation target in blocks (1 - 1008)"},
-                    {"threshold", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "", "The proportion of transactions in a given feerate range that must have been\n"
+                    {"threshold", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "0.95", "The proportion of transactions in a given feerate range that must have been\n"
```
Comment (jonasschnelli) [inline src/rpc/mining.cpp]: what about using a stringprint of the `threshold` variable?

**Comment by jonasschnelli** [comment]
utACK fa0c24c96e9937f666dcdd83d12145720c7b0329
Somehow I would be in favour of putting brackets around descriptive values (`(null)` instead of `null` and `(fallback to nodeid)` instead of `fallback to nodeid`) to avoid confusion between real default values and behavioural description.

**Comment by kristapsk** [comment]
ACK fa0c24c96e9937f666dcdd83d12145720c7b0329

File: src/rpc/net.cpp [hunk 5]
```diff
@@ -518,8 +518,8 @@ static UniValue setban(const JSONRPCRequest& request)
...
                     {"command", RPCArg::Type::STR, /* opt */ false, /* default_val */ "", "'add' to add an IP/Subnet to the list, 'remove' to remove an IP/Subnet from the list"},
-                    {"bantime", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "", "time in seconds how long (or until when if [absolute] is set) the IP is banned (0 or empty means using the default time of 24h which can also be overwritten by the -bantime startup argument)"},
-                    {"absolute", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "", "If set, the bantime must be an absolute timestamp in seconds since epoch (Jan 1 1970 GMT)"},
+                    {"bantime", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "0", "time in seconds how long (or until when if [absolute] is set) the IP is banned (0 or empty means using the default time of 24h which can also be overwritten by the -bantime startup argument)"},
+                    {"absolute", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "false", "If set, the bantime must be an absolute timestamp in seconds since epoch (Jan 1 1970 GMT)"},
```
Comment (ryanofsky) [inline src/rpc/net.cpp]: /If set/If true/ might be more accurate

**Review by ryanofsky — APPROVED** [review]
ACK fa0c24c96e9937f666dcdd83d12145720c7b0329. I checked help output using the hack from https://github.com/bitcoin/bitcoin/pull/14726#pullrequestreview-175423309

This is a nice documentation change. If you wanted to follow up with more improvements I have some suggestions below:

- The fallback descriptions like "`fallback to wallet's default`" are vague and could be described in more detail.
- I don't think saying `(default=null)` or `(default=omitted)` is ever helpful. Just because of the way we check arguments, omitting an argument is always equivalent to passing `null`. And `(default=omitted)` is just a tautology. The point of the default documentation should be to say what an RPC call will do when  an argument is `null` or omitted. I actually think it would be better to drop the all the `(default=null)` and `(default=omitted)` strings if they can't be replaced with something more constructive, to avoid creating a bad precedent for future changes.

**Comment by kristapsk** [comment]
> I don't think saying (default=null) or (default=omitted) is ever helpful. (...) I actually think it would be better to drop the all the (default=null) and (default=omitted) strings if they can't be replaced with something more constructive, to avoid creating a bad precedent for future changes.

If there is "default=" string outputted for each optional argument, functional test could be made that checks that people don't forget to document optional argument defaults for a new RPCs.

**Comment by kristapsk** [comment]
Oh, yes, short list of a few exceptions probably is better idea.

**Comment by MarcoFalke** [comment]
I have changes piled up locally to enforce that it is either a required arg or a default value is provided at compile time.

**Comment by ryanofsky** [comment]
@MarcoFalke what do you want to do with all the `(default=null)` / `(default=omitted)` cruft? Do you want to address it in a future PR or maybe mark this one up for grabs so someone else can fill in useful default values?

**Comment by MarcoFalke** [comment]
@ryanofsky Your suggestion is a one-line change after #14918, so either you leave feedback there or create a pull request after #14918 is merged.

**Pull request merged**